rx-gen module can't be used in OSGI environment crashing rx-java2 bundle

After refactoring for 3.6.0 version (see [link](https://github.com/vert-x3/vertx-rx/commit/01a41fde440bd5609c9483633e65a0f9e696cba0#diff-78bfa76ace3017edfba568e218c46f6a])) rx-java2 bundle can't be started in OSGI environment due to unresolved dependencies to rx-gen bundle (which is not bundle due to missing OSGI manifest):
```
ERROR: Bundle com.github.timofeevda.jstressy-vertx [30] Error starting file:/vagrant/jstressy/jstressy-assembly-osgi/target/bundles/application/com.github.timofeevda.jstressy.jstressy-vertx_1.0.10-SNAPSHOT.jar (org.osgi.framework.BundleException: Activator start error in bundle com.github.timofeevda.jstressy-vertx [30].)
java.lang.NoClassDefFoundError: io/vertx/lang/rx/TypeArg
        at io.vertx.reactivex.core.Vertx.<clinit>(Vertx.java:90)
        at com.github.timofeevda.jstressy.vertx.StressyVertxService.<init>(StressyVertxService.kt:36)
        at com.github.timofeevda.jstressy.vertx.internal.Activator.start(Activator.kt:39)
        at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697)
        at org.apache.felix.framework.Felix.activateBundle(Felix.java:2240)
        at org.apache.felix.framework.Felix.startBundle(Felix.java:2146)
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1373)
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: io.vertx.lang.rx.TypeArg not found by vertx-rx-java2 [27]
        at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1639)
        at org.apache.felix.framework.BundleWiringImpl.access$200(BundleWiringImpl.java:80)
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2053)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 9 more
```
As a workaround you can wrap rx-gen as OSGI bundle, but would be great to have it out of the box and use it in OSGI framework as is.

Well, I think I need some clarifications here, because this is the first time I'm trying to make something in Vertx repositories.

After I fixed rx-gen build and made it as an OSGI bundle with correct MANIFEST vertx-codegen started to be required due to transitive dependency.

#### First issue
This is a log from Apache Felix framework when I'm trying to run my app with vertx-codegen in bundles list:
```
ERROR: Bundle vertx-codegen [46] Error starting file:/C:/Sources/vagrant-gwtrxjs/jstressy/jstressy-assembly-osgi/target/bundles/application/vertx-codegen-3.6.1-SNAPSHOT.jar (org.osgi.framework.BundleException: Unable to resolve vertx-codegen [46](R 46.0): missing requirement [vertx-codegen [46](R 46.0)] osgi.wiring.package; (osgi.wiring.package=com.sun.source.tree) Unresolved requirements: [[vertx-codegen [46](R 46.0)] osgi.wiring.package; (osgi.wiring.package=com.sun.source.tree)])
org.osgi.framework.BundleException: Unable to resolve vertx-codegen [46](R 46.0): missing requirement [vertx-codegen [46](R 46.0)] osgi.wiring.package; (osgi.wiring.package=com.sun.source.tree) Unresolved requirements: [[vertx-codegen [46](R 46.0)] osgi.wiring.package; (osgi.wiring.package=com.sun.source.tree)]
        at org.apache.felix.framework.Felix.resolveBundleRevision(Felix.java:4149)
        at org.apache.felix.framework.Felix.startBundle(Felix.java:2119)
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1373)
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
        at java.lang.Thread.run(Unknown Source)
```
After I fixed this I got the following error (the reason is that vertx-codegen doesn't explicitly require vertx-core and required classes are excluded from build)
```
Caused by: java.lang.ClassNotFoundException: *** Package 'io.vertx.core' is imported by bundle vertx-codegen [46] from bundle vertx-codegen [46], but the exported package from bundle vertx-codegen [46] does not contain the requested class 'io.vertx.core.Handler'. Please verify that the class name is correct in the importing bundle vertx-codegen [46] and/or that the exported package is correctly bundled in vertx-codegen [46]. ***
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2066)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at org.apache.felix.framework.BundleWiringImpl.getClassByDelegation(BundleWiringImpl.java:1414)
        at org.apache.felix.framework.BundleWiringImpl.searchImports(BundleWiringImpl.java:1660)
        at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1590)
        at org.apache.felix.framework.BundleWiringImpl.access$200(BundleWiringImpl.java:80)
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2053)
        ... 9 more
Caused by: java.lang.ClassNotFoundException: io.vertx.core.Handler not found by vertx-codegen [46]
        at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1639)
        at org.apache.felix.framework.BundleWiringImpl.access$200(BundleWiringImpl.java:80)
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2053)
        ... 15 more
```

I've fixed these two issues with the following changes in vertx-codegen pom.xml from:
``` 
<configuration>
          <bnd><![CDATA[
          Import-Package: \
                org.mvel2*;resolution:=optional, \
                io.vertx.core, \
                io.vertx.core.buffer, \
                  *
                -exportcontents: !*impl, !examples, *
          ]]></bnd>
        </configuration>
```
to
```
<configuration>
          <bnd><![CDATA[
          Import-Package: \
                org.mvel2*;resolution:=optional, \
                io.vertx.core, \
                io.vertx.core.buffer, \
                  *;resolution:=optional
                -exportcontents: !*impl, !examples, !io.vertx.core*,*
          ]]></bnd>
        </configuration>
```
After these changes Felix stops blaming on vertx-codegen.

#### Second issue
The next part is rx-gen and rx-java-gen/rx-java2-gen. Am I correct to say that vertx-rx-java2-gen and vertx-rx-java-gen are not needed in runtime actually? I made them provided so that I don't get them in classpath or in required OSGI bundles list

Then I made dependencies to rxjava and rxjava2 explicit in vertx-rx-java and vertx-rx-java2. This also requires changes in bundle's Import-Package definition. 

After all these changes I was able to run bundles (vertx-codege, vertx-rx-gen, vertx-rx-java2) in Apache Felix without any issues.

Making rx-java-gen/rx-java2-gen provided fixes #173 because it excludes these modules from the list of dependencies.

If I'm on the right way I can make PRs to vertx-codegen and vertx-rx repositories for further review and discussion.

--- a/rx-gen/pom.xml
+++ b/rx-gen/pom.xml
@@ -16,6 +16,14 @@
   <name>Vert.x Rx Code Generator</name>
 
   <properties>
+    <!-- The path to the assembly jar descriptor -->
+    <rxgen.descriptor>${basedir}/src/main/assembly/artifact.xml</rxgen.descriptor>
+    <asciidoc.dir>${project.basedir}/src/main/asciidoc</asciidoc.dir>
+    <lang>java</lang>
+    <!-- There is an issue with source plugin because we don't get the generated sources
+         because of BSC plugin, also the source is messy because we keep a version of vert.x core
+         so better use the assembly plugin instead -->
+    <source.skip>true</source.skip>
   </properties>
 
   <dependencies>
@@ -40,5 +48,61 @@
       <scope>test</scope>
     </dependency>
   </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>biz.aQute.bnd</groupId>
+        <artifactId>bnd-maven-plugin</artifactId>
+        <version>3.2.0</version>
+        <executions>
+          <execution>
+            <id>default-bnd-process</id>
+            <goals>
+              <goal>bnd-process</goal>
+            </goals>
+          </execution>
+        </executions>
+        <configuration>
+          <bnd><![CDATA[
+          Import-Package: \
+              io.vertx.codegen,\
+              io.vertx.codegen.annotations,\
+              io.vertx.codegen.doc,\
+              io.vertx.codegen.type,\
+              javax.lang.model.element,\
+              *;resolution:=optional
+          Export-Package: \
+              io.vertx.lang.rx*
+          ]]></bnd>
+        </configuration>
+      </plugin>
+
+      <plugin>
+        <artifactId>maven-assembly-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>package-jar</id>
+            <phase>package</phase>
+            <goals>
+              <goal>single</goal>
+            </goals>
+            <configuration>
+              <attach>true</attach>
+              <appendAssemblyId>false</appendAssemblyId>
+              <descriptors>
+                <descriptor>${rxgen.descriptor}</descriptor>
+              </descriptors>
+              <archive>
+                <manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>
+              </archive>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+
+
+    </plugins>
+  </build>
 
 </project>

--- a/rx-gen/src/main/asciidoc/vertx-rx/java/index.adoc
+++ b/rx-gen/src/main/asciidoc/vertx-rx/java/index.adoc
@@ -1,0 +1,2 @@
+= Common generator API for RxJava/RxJava2
+:toc: left

--- a/rx-gen/src/main/assembly/artifact.xml
+++ b/rx-gen/src/main/assembly/artifact.xml
@@ -1,0 +1,18 @@
+<assembly xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.1"
+          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+          xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.1 http://maven.apache.org/xsd/assembly-1.1.1.xsd">
+  <id>artifact</id>
+  <formats>
+    <format>jar</format>
+  </formats>
+  <includeBaseDirectory>false</includeBaseDirectory>
+  <fileSets>
+    <fileSet>
+      <directory>${project.build.outputDirectory}</directory>
+      <outputDirectory>/</outputDirectory>
+      <includes>
+        <include>io/vertx/lang/rx/**</include>
+      </includes>
+    </fileSet>
+  </fileSets>
+</assembly>

--- a/rx-java/pom.xml
+++ b/rx-java/pom.xml
@@ -30,6 +30,11 @@
 
   <dependencies>
 
+    <dependency>
+      <groupId>io.reactivex</groupId>
+      <artifactId>rxjava</artifactId>
+    </dependency>
+
     <!-- Stack dependencies -->
     <dependency>
       <groupId>io.vertx</groupId>
@@ -324,6 +329,7 @@
       <groupId>io.vertx</groupId>
       <artifactId>vertx-rx-java-gen</artifactId>
       <version>3.6.1-SNAPSHOT</version>
+      <scope>provided</scope>
     </dependency>
 
     <!-- Test dependencies -->
@@ -660,6 +666,9 @@
               io.vertx.core.spi.logging,\
               io.vertx.core.spi.metrics,\
               io.vertx.core.streams,\
+              rx,\
+              rx.functions,\
+              rx.plugins,\
               *;resolution:=optional
           Export-Package: \
               io.vertx.rxjava*,\

--- a/rx-java2/pom.xml
+++ b/rx-java2/pom.xml
@@ -29,6 +29,15 @@
 
   <dependencies>
 
+    <dependency>
+      <groupId>io.reactivex.rxjava2</groupId>
+      <artifactId>rxjava</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.reactivestreams</groupId>
+      <artifactId>reactive-streams</artifactId>
+    </dependency>
+
     <!-- Stack dependencies -->
     <dependency>
       <groupId>io.vertx</groupId>
@@ -323,6 +332,7 @@
       <groupId>io.vertx</groupId>
       <artifactId>vertx-rx-java2-gen</artifactId>
       <version>3.6.1-SNAPSHOT</version>
+      <scope>provided</scope>
     </dependency>
 
     <!-- Test dependencies -->
@@ -445,6 +455,11 @@
               io.vertx.core.spi.logging,\
               io.vertx.core.spi.metrics,\
               io.vertx.core.streams,\
+              io.reactivex,\
+              io.reactivex.disposables,\
+              io.reactivex.functions,\
+              io.reactivex.internal.subscriptions,\
+              io.reactivex.plugins,\
               *;resolution:=optional
           Export-Package: \
               io.vertx.reactivex*,\

